v2 CLI - initial train and predict functionality

kevingreenman opened the pull request:
In progress. The following currently runs to completion without errors:

```
chemprop train -i tests/data/regression.csv
```

davidegraff commented on chemprop/v2/models/modules/readout.py:
```diff
@@ -104,6 +107,7 @@ def train_step(self, Z: Tensor) -> Tensor:
 class MveFFN(RegressionFFN):
     n_targets = 2
     _default_criterion = loss.MVELoss()
+    _default_metric = metrics.MSEMetric()
```
don't need to redefine because they all inherit from `RegressionFFN`

davidegraff commented on chemprop/v2/cli/utils/config.py:
```diff
@@ -2,5 +2,6 @@
 import os
 from pathlib import Path
 
+CHK_DIR = Path(os.getenv("CHEMPROP_CHK_DIR", "model_checkpoints"))
```
```suggestion
CKPT_DIR = Path(os.getenv("CHEMPROP_CKPT_DIR", "model_checkpoints"))
```

davidegraff commented on chemprop/v2/cli/utils/config.py:
can we rename this to `conf.py`. Seems to be a more common convention for package-level constants.

davidegraff commented on chemprop/v2/cli/utils/utils.py:
```diff
@@ -75,3 +75,14 @@ def validate_loss_function(
+    ) -> list:
+    if columns is None:
+        return None
+    if type(columns[0]) == str:
+        columns = [i for i,e in enumerate(header) if e in columns]
+    assert type(columns[0]) == int
```
```suggestion
    if all(isinstance(col, str) for col in columns):
        columns = [i for i, name in enumerate(header) if name in columns]
    if not all(isinstance(col, int) for col in columns):
        raise ValueError #TODO: add message
```
I'm assuming that `columns` comes in as one of  `None | list[str], list[int]` and if it's `list[str]` then it gets converted to `list[int]`?

davidegraff commented on chemprop/v2/cli/train.py:
```diff
@@ -156,7 +154,14 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
+            "selu",
+            "elu",
+        ],  # TODO: should these be lowercase?
         help="activation function in message passing/FFN layers",
```
activations are controlled by the `chemprop.v2.models.utils.Activation` enum

kevingreenman replied:
so this means all lowercase is correct, right? or does something need to change?

hwpang replied:
You can replace the choices with

```
list(Activation.keys())
```

to get all available ones.

davidegraff replied:
yeah just change this to
```python
choices=Activation.keys()
type=lambda x: x.lower()
```
The `type` arg is just a lambda that automatically puts all strings into lowercase so the input is case-insensitive.

davidegraff commented on chemprop/v2/cli/train.py:
```diff
@@ -247,18 +255,20 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
+        default="regression",
+        action=RegistryAction(
+            ReadoutRegistry
+        ),  # TODO: is this correct? The choices should be ['regression', 'classification', 'multiclass', 'spectra']
```
I think we might want to change this over to `--task-type` because a `ReadoutFFN` controls the prediction task, which itself implicitly defines the dataset type. That is, there is no internal concept of a dataset type anymore

davidegraff commented on chemprop/v2/cli/train.py:
```diff
@@ -341,33 +355,29 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
         "--evidential-regularization",
-        type=float, 
+        type=float,
         default=0.2,  # TODO: the default in v1 was 0. Do we want it to default to 0.2 in v2?
```
probably best to keep it the same to eliminate any surprises. In principle, however, it would be better to stick with the paper default

kevingreenman replied:
see #352 for explanation of why 0.0 was chosen as the default in v1 instead of 0.2 from the paper. I think we should stick with 0.0.

davidegraff commented on chemprop/v2/cli/train.py:
```diff
@@ -433,14 +423,17 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
+        choices=[
+            "random",
+            "scaffold",
+        ],  #'scaffold_balanced', 'predetermined', 'crossval', 'cv', 'cv-no-test', 'index_predetermined', 'random_with_repeated_smiles'],
```
```suggestion
        ],  #'scaffold_balanced', 'predetermined', 'crossval', 'cv', 'cv-no-test', 'index_predetermined', 'random_with_repeated_smiles'], # TODO: make data splitting CLI play nicely with astartes backend
```

davidegraff commented on chemprop/v2/cli/train.py:
```diff
@@ -433,14 +423,17 @@ def add_train_args(parser: ArgumentParser) -> ArgumentParser:
     )
     split_args.add_argument(
         "--split-sizes",
         type=list[float],
```
```suggestion
        type=float,
        nargs=3,
```

hwpang commented:
With my changes, now `chemprop predict -i tests/data/regression_test_smiles.csv --checkpoint-path /path/to/checkpoint.ckpt` works.

kevingreenman commented:
Fixes #410 with `**kwargs` in `MolGraphDataLoader`

hwpang commented:
@kevingreenman Thanks for linking the issue! I actually added `**kwargs` to dataloader in my merged PR #471 (https://github.com/chemprop/chemprop/pull/471/commits/2cdf83cfe8b8be100e7e8f3734033a57fc6d15fd). I can rebase this branch with `v2/dev` if that's ok with you.

kevingreenman commented:
@hwpang thanks for pointing that out, I had missed that detail in #471. I've closed that issue now. I have a couple minor changes I'm making to prediction now, then I can do the rebase and merge

davidegraff commented on chemprop/v2/cli/predict.py:
```diff
@@ -58,7 +59,7 @@ def add_predict_args(parser: ArgumentParser) -> ArgumentParser:
         "--output",
         "--preds-path",
         type=str,
```
```suggestion
```

kevingreenman replied:
why?

davidegraff replied:
args are always parsed as strings, so this doesn't do anything

davidegraff commented on chemprop/v2/cli/predict.py:
```diff
@@ -153,32 +154,29 @@ def add_predict_args(parser: ArgumentParser) -> ArgumentParser:
 def validate_predict_args(args):
-    pass
+    # TODO: onces args.checkpoint_dir and args.checkpoint are consolidated, need to change this as well. Not able to make this required in common.py as it may not be provided for training.
+    assert args.checkpoint_path is not None, "Must provide a checkpoint path for prediction."
```
replace with `raise ValueError`

davidegraff commented on chemprop/v2/cli/predict.py:
```diff
@@ -199,81 +197,61 @@ def main(args):
+
+    # TODO: might want to write a shared function for this as train.py might also want to do this.
+    df_test = pd.read_csv(args.test_path)
+    preds = np.vstack(predss)
```
```suggestion
    preds = torch.concat(predss, 1).numpy()
```

davidegraff commented on chemprop/v2/cli/predict.py:
```diff
@@ -199,81 +197,61 @@ def main(args):
     logger.info(model)
 
     trainer = pl.Trainer(
         logger=None,
         enable_progress_bar=True,
         accelerator="auto",
         devices=args.n_gpu if torch.cuda.is_available() else 1,
-        max_epochs=args.epochs,
     )
```
wrap this whole block using [`torch.inference_mode()`](https://pytorch.org/docs/stable/generated/torch.inference_mode.html)

kevingreenman replied:
@davidegraff is the right way to do this by putting the call to `trainer.predict` in a `with torch.inference_mode():` block (as [here](https://github.com/Lightning-AI/lightning/issues/11530#issuecomment-1067566245)), or to add a `inference_mode=False` argument to `pl.Trainer` (from the [documentation](https://lightning.ai/docs/pytorch/latest/common/trainer.html))?

I'm admittedly a bit confused by the documentation, since it says `Trainer(inference_mode=True)` is the default and `Trainer(inference_mode=False)` is equivalent to `torch.no_grad`, while I would expect the opposite to be true, unless I'm misunderstanding something...

davidegraff replied:
yeah that's a good question. Looking at the source, it doesn't seem like the argument actually does _anything_. Even if it does do something in the future, it won't hurt us if our code is already inside a `with`-block (i.e., the PyTorch docs)

hwpang replied:
@kevingreenman By reading the documentation, I think `Trainer(inference_mode=True)` means that `torch.inference_mode():` is used by Trainer internally during model evaluation (i.e., when `Trainer.validate`, `Trainer.test`, and `Trainer.predict` are called), while `Trainer(inference_mode=False)` uses `torch.no_grad():` during model evaluation. But I would like to know @davidegraff's thoughts as well.

davidegraff reviewed (commented):
just double checking (i should know since i wrote the code) that prediction of regression targets automatically applies the output transform (i.e., multiplies by `scale` and shifts by `loc`) and that binary classification outputs are passed to `nn.Sigmoid()`?

hwpang commented:
> just double checking (i should know since i wrote the code) that prediction of regression targets automatically applies the output transform (i.e., multiplies by `scale` and shifts by `loc`) and that binary classification outputs are passed to `nn.Sigmoid()`?

@davidegraff Thanks for double checking! For regression, the readout layer applies them in forward, see [here](https://github.com/chemprop/chemprop/blob/v2/dev/chemprop/v2/models/modules/readout.py#L97) for an example. For classification, the `Sigmoid` step is done in forward as well, see [here](https://github.com/chemprop/chemprop/blob/v2/dev/chemprop/v2/models/modules/readout.py#L162).

kevingreenman commented:
@davidegraff @hwpang I have addressed most of the comments and resolved merge conflicts. I confirmed the following commands run as expected:
```
chemprop train -i tests/data/regression.csv --epochs 1
chemprop predict -i tests/data/regression_test_smiles.csv -o test_preds.csv --checkpoint-path /path/to/checkpoint.ckpt
```

there are just 3 comments I'd like @davidegraff's feedback on before merging

KnathanM commented on chemprop/v2/cli/predict.py:
```diff
@@ -39,160 +40,156 @@ def add_args(cls, parser: ArgumentParser) -> ArgumentParser:
         type=str,
-        help="Path to CSV or PICKLE file where predictions will be saved.",
+        required=True,
+        help="Path to CSV or PICKLE file where predictions will be saved. If the file extension is .pkl, will be saved as a PICKLE file. If not provided and the test_path is /path/to/test/test.csv, predictions will be saved to /path/to/test/test_preds.csv.",
```
It doesn't make sense to require the user to give this argument and also tell them in the help that we have a default if they don't provide it. One of those should be removed.

davidegraff commented:
> @davidegraff @hwpang I have addressed most of the comments and resolved merge conflicts. I confirmed the following commands run as expected:
> there are just 3 comments I'd like @davidegraff's feedback on before merging

I only saw two comments? but otherwise they should be quick changes and then we're good to go.

hwpang commented on chemprop/v2/cli/train.py:
```diff
@@ -799,16 +608,17 @@ def main(args):
         )
         criterion = readout_cls._default_criterion
 
+    print(readout_cls)
```
Can you remove this?
